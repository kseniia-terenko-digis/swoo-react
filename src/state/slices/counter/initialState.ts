import { ICounterState } from './interfaces/iCounterState';

export const initialState: ICounterState = {
  value: 0,
};
