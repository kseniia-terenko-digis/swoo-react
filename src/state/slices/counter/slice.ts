import { createSlice } from '@reduxjs/toolkit';
import { initialState } from './initialState';

export const counterSlice = createSlice({
  name: 'counter',
  initialState,
  reducers: {
    increment: (state) => {
      state.value += 1;
    },
    decrement: (state) => {
      state.value -= 1;
    },
  },
});
