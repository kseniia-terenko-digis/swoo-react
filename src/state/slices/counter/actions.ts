import { counterSlice } from './slice';

export const { increment, decrement } = counterSlice.actions;
