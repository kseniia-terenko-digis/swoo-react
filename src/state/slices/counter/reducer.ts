import { counterSlice } from './slice';

export default counterSlice.reducer;
