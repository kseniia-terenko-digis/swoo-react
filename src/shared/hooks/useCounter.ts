import { useCallback } from 'react';
import { useAppDispatch } from '../../state/hooks/useAppDispatch';
import { useAppSelector } from '../../state/hooks/useAppSelector';
import { decrement, increment } from '../../state/slices/counter/actions';

export const useCounter = () => {
  const counterValue = useAppSelector((state) => state.counter.value);
  const dispatch = useAppDispatch();

  const incrementCounter = useCallback(() => dispatch(increment()), [dispatch]);
  const decrementCounter = useCallback(() => dispatch(decrement()), [dispatch]);

  return { counterValue, incrementCounter, decrementCounter };
};
