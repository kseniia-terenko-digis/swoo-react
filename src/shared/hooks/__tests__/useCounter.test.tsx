import { useCounter } from '../useCounter';
import { renderHook } from '@testing-library/react-hooks';
import { decrement, increment } from '../../../state/slices/counter/actions';

const mockDispatch = jest.fn();

jest.mock('../../../state/hooks/useAppSelector', () => ({
  useAppSelector: () => 0,
}));

jest.mock('../../../state/hooks/useAppDispatch', () => ({
  useAppDispatch: () => mockDispatch,
}));

describe('useQueryParamState hook', () => {
  const { result } = renderHook(() => useCounter());

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('result matches snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('decrementCounter and fire dispatch with correct parameters', () => {
    result.current.decrementCounter();

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(mockDispatch).toHaveBeenCalledWith(decrement());
  });

  it('incrementCounter and fire fire dispatch with correct parameters', () => {
    result.current.incrementCounter();

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(mockDispatch).toHaveBeenCalledWith(increment());
  });
});
