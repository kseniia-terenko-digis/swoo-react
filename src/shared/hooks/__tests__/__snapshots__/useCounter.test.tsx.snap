// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useQueryParamState hook result matches snapshot 1`] = `
Object {
  "all": Array [
    Object {
      "counterValue": 0,
      "decrementCounter": [Function],
      "incrementCounter": [Function],
    },
  ],
  "current": Object {
    "counterValue": 0,
    "decrementCounter": [Function],
    "incrementCounter": [Function],
  },
  "error": undefined,
}
`;
